db/document.go

```diff
@@ -38,6 +39,171 @@ const (
 	DocUnmarshalNone                                     // No unmarshalling (skips import/upgrade check)
 )
 
+// Struct populated by an incoming document
+type IncomingDocument struct {
+	_body   Body
+	RawBody []byte
+	SpecialProperties
+}
+
+// Stores special properties which are a part of an incoming document
+type SpecialProperties struct {
+	DocID       string
+	RevID       string
+	Deleted     bool
+	Expiry      uint32
+	Attachments AttachmentsMeta
+	Revisions   Revisions
+}
+
+func NewIncomingDocument(docID string, body []byte) *IncomingDocument {
+	return &IncomingDocument{
+		RawBody: body,
+		SpecialProperties: SpecialProperties{
+			DocID: docID,
+		},
+	}
+}
+
+func (doc *IncomingDocument) GetSyncFnBody() (map[string]interface{}, error) {
+	//if doc._body != nil {
+	//	return doc._body, nil
+	//}
+
+	var syncFnBody map[string]interface{}
+	buf := bytes.NewBuffer(doc.RawBody)
+	d := json.NewDecoder(buf)
+	d.UseNumber()
+	err := d.Decode(&syncFnBody)
+	if err != nil {
+		return nil, err
+	}
+
+	stampSpecialProperties(syncFnBody, doc.SpecialProperties)
+
+	doc._body = syncFnBody
+	return doc._body.Copy(BodyDeepCopy), nil
+}
+
+func (doc *IncomingDocument) GetMutableBody() Body {
+	// we can avoid a deep copy by just unmarshalling raw bytes, if available
+	if doc.RawBody != nil {
+		var b Body
+		err := b.Unmarshal(doc.RawBody)
+		if err == nil {
+			return b
+		}
+		// Error unmarshalling raw body, try to use the existing _body
+		base.Warnf(base.KeyAll, "Unable to unmarshal document body from raw body : %s", err)
+	}
+
+	// We didn't have raw bytes available, but if we do have a body to copy
+	if doc._body != nil {
+		// need to deep copy so callers can mutate
+		return doc._body.Copy(BodyDeepCopy)
+	}
+
+	return nil
+}
+
+func (doc *IncomingDocument) CreateRevID(generation int, parentRevID string) (string, error) {
+	var kvPairsToAdd []base.KVPair
+
+	// if doc.Attachments != nil {
+	// 	kvPairsToAdd = append(kvPairsToAdd, base.KVPair{Key: BodyAttachments, Val: doc.Attachments})
```

Comment:
Previously it looks like we were including attachments when generating a rev ID, how come this is commented?